### PR TITLE
Add noDamage Command

### PR DESCRIPTION
**Pull request by Weilin1992**
### Contains
add god command to immune to all damage mentioned in #2837

### How to test

Type the command "noDamage", and you will be immutable to all damage. Type "GodMode" again to remove God mode, need cheat permission

### Notes
This is a simple command make player immutable to all types of damage. Later I will submit another PR to add another command to make player immutable to some particular type of damage

### CODE REVIEW

**Comment by GooeyHub** [comment]
Can one of the admins please verify this patch?

**Comment by OvermindDL1** [comment]
`nodamage` or `noDamage` might be a better command name?

Also instead of adding a new field to the health component, making it larger, what about some kind of event hook to cancel damage applied to the health component, or better yet, remove the health component entirely?

**Comment by Weilin1992** [comment]
@OvermindDL1 I don't think remove health component is better, since we will lost some information, such as current health, current max health. If the component can be disable like unity3d, that will be a good choice.But If we want to add a disable property it will make all component larger.

I have another idea, how about add a noDamage component. check if the entity have this noDamage component, we can use a command to remove or add this component. This will not make health component  larger.Further we can develop a buff system based on such component.

**Comment by OvermindDL1** [comment]
> @OvermindDL1 I don't think remove health component is better, since we will lost some information, such as current health, current max health. If the component can be disable like unity3d, that will be a good choice.But If we want to add a disable property it will make all component larger.

You can always make a new 'DisabledHealthComponent' just to store the information for recovery later.

> I have another idea, how about add a noDamage component. check if the entity have this noDamage component, we can use a command to remove or add this component. This will not make health component larger.Further we can develop a buff system based on such component.

This could work, as long as all the systems that affect the HealthComponent checked it too, but two issues I see is lots of 'if's around for that and not everything will check it, hence why I think it is better to just remove the health component at all, can store the information in a dummy component instead if it is really wanted to be stored (not all situations would need that, though would be useful for this specific command).

**Comment by Weilin1992** [comment]
@OvermindDL1 I change the code as you suggested. remove HealthComponent and store the information to DisabledHealthComponent.

**Comment by OvermindDL1** [comment]
I'm not the Authors, just someone that programs on Terasology at times so I know how I would prefer to use these changes.  ^.^

It looks good though, should be quite efficient, reducing the number of checks, and can be used for more than just players with ease.

And idea though, I wonder if the disabled component could just subclass the normal health component or if the internal ECS would identify them the same...  If this were a static language I'd say just make the healthcomponent an internal field (not a pointer), and that might still be useful here, hmm...

I wonder if it is worth making a `DisabledComponentsComponent`, that is just a list of components that are 'disabled' but you want to hold on to later.  This would allow it to be far more generic and useful for any component where you want to keep the state when you put it back on but still want it 'gone' for now...  Ideas anyone?

File: modules/Core/src/main/java/org/terasology/logic/health/HealthComponent.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 /**
  */
-public final class HealthComponent implements Component {
```
Comment (OvermindDL1) [inline modules/Core/src/main/java/org/terasology/logic/health/HealthComponent.java]: Hmm, making this non-final could have a performance issue...  Java needs a way to use the same memory as a different type as is easy in C++, or to not use individual components, hmm...

**Comment by oniatus** [comment]
What's the state of this PR?

Most recent discussion about disabling components would be #2896 which seems to favor a separate component to mark an entity as immune.

**Comment by Weilin1992** [comment]
@oniatus I will work on this issue #2896 recently. Maybe this pr can be closed because i have to rewrite my code

**Comment by Cervator** [comment]
Good to hear from you @Weilin1992 !

Yep probably easier to rewrite. Will close this. Hoping to see more from you in the future :-)

**Pull request closed without merging**